### ISSUE TITLE
[Bug]: copy_to_directory fails with host!=exec RBE

### ISSUE THREAD

**Issue by anguslees** [aspect-build/bazel-lib#466]
### What happened?

I'm using bazel on a darwin/arm64 host, pointing at a linux/amd64 RBE cluster.  `copy_to_directory` fails with "cannot execute binary file" because bazel is trying to run a linux/amd64 binary locally.

Here is one example failure, where `copy_to_directory` is invoked as part of `rules_oci`'s `oci_pull`
```
ERROR: /private/var/tmp/_bazel_gus/798cdab9ca916cc0b6f69f7c876590eb/external/distroless_static_linux_arm64/BUILD.bazel:47:18: Copying files to directory distroless_static_linux_arm64/blobs/sha256 failed: (Exit 126): copy_to_directory failed: error executing command (from target @distroless_static_linux_arm64//:blobs)
  (cd /private/var/tmp/_bazel_gus/798cdab9ca916cc0b6f69f7c876590eb/execroot/com_canva_infrastructure && \
  exec env - \
  external/copy_to_directory_linux_amd64/copy_to_directory bazel-out/darwin_arm64-fastbuild/bin/external/distroless_static_linux_arm64/blobs_config.json)
# Configuration: 552dda68697b5bb17c41b4bef2af5919b102e5481624a5bb19489b4de2a07a0c
# Execution platform: //tools/build/bazel/toolchains/remote:ubuntu-act-22-04-platform
external/copy_to_directory_linux_amd64/copy_to_directory: external/copy_to_directory_linux_amd64/copy_to_directory: cannot execute binary file
```

Note bazel _thinks_ the execution platform is my RBE platform (ubuntu-act-22-04-platform), and has selected `copy_to_directory_linux_amd64` appropriately, but the action is _actually_ executed locally on my darwin/amd64 mac and fails ("Exit 126" and "cannot execute binary file").

After much head scratching, I discovered `@aspect_bazel_lib//lib/private/copy_common.bzl` `COPY_EXECUTION_REQUIREMENTS`, which forces copy commands to be performed locally.  This effectively discards all the toolchain resolution hard work, and results in the above error.

### Version

Development (host) and target OS/architectures:

host = darwin/arm64
exec = linux/amd64
target = linux/amd64

Output of `bazel --version`:

bazel 6.1.2

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file:

Aspect rules v1.32.1

Language(s) and/or frameworks involved:

### Any other information?

I think the fix is either:

- Remove all of COPY_EXECUTION_REQUIREMENTS.  This is what I've done locally.  I appreciate the optimisation goal, but these are dubious at best with a sufficiently large+fast RBE cache, remote asset API, and a slow network link from my client to RBE cluster.  If performance _is_ an issue, it's certainly easier in my case to expand the size of CAS disk rather than expand any other part of the system.  I have not confirmed if the caveat about src treeartifacts not working over remote-exec api still applies, but it hasn't affected my use cases in ways that I've noticed .. yet.
- Not use toolchain resolution for copy_to_directory.  If we're forcing it to execute locally, then we also want to force it to use the host platform's copy_to_directory executable.
- Lean further into local execution, and add exec_compatible_with=HOST_CONSTRAINTS on all relevant rules (or actions, with exec_groups)

**Comment by alexeagle** [comment]
Thanks for the detailed explanation. I think I agree with your first fix, to remove our "always run the action locally" optimization, making that an opt-in instead.
@gregmagolan does that sound right to you?

**Comment by alexeagle** [comment]
https://github.com/bazelbuild/bazel-skylib/blob/0171c69e5cc691e2d0cd9f3f3e4c3bf112370ca2/rules/private/copy_common.bzl is probably correct. @anguslees could you confirm if that fixes your problem as well?

**Comment by pullard** [comment]
I'm still experiencing this same issue when running RBE on a darwin/arm64 host, pointing at a linux/amd64 RBE cluster with the latest `1.34.1` release:

```
external/copy_directory_linux_amd64/copy_directory: external/copy_directory_linux_amd64/copy_directory: cannot execute binary file
```

Removing the `COPY_EXECUTION_REQUIREMENTS` is the only thing that has "worked". Although, I do encounter another blocker and similar failure invoking `bazel` managed `node` which incorrectly attempts to invoke the `linux/amd64` binary over RBE.

**Comment by anguslees** [comment]
Apologies for not testing the patch before # merge.  I think `no-remote` is still going to influence spawn strategy, and thus fight against regular platform selection.

**Comment by alexeagle** [comment]
Okay then this bug also exists in bazel-skylib, maybe we should raise it there as well and get some Googler input on it.

**Comment by alexeagle** [comment]
We'll add a command-line flag to turn off our automatic addition of execution requirements.

**Comment by alexeagle** [comment]
#604 tracks that change

**Comment by alexeagle** [comment]
Fixed by #606 which gives a custom bazel flag to turn off our execution requirements defaults altogether.
